server/services/store/sqlstore/board_insights.go

```diff
@@ -0,0 +1,147 @@
+package sqlstore
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/mattermost/focalboard/server/model"
+
+	sq "github.com/Masterminds/squirrel"
+	mmModel "github.com/mattermost/mattermost-server/v6/model"
+
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+func (s *SQLStore) getTeamBoardsInsights(db sq.BaseRunner, teamID string, userID string, since int64, offset int, limit int, boardIDs []string) (*model.BoardInsightsList, error) {
+	boardsHistoryQuery := s.getQueryBuilder(db).
+		Select("boards.id, boards.icon, boards.title, count(boards_history.id) as count, boards_history.modified_by, boards.created_by").
+		From(s.tablePrefix + "boards_history as boards_history").
+		Join(s.tablePrefix + "boards as boards on boards_history.id = boards.id").
+		Where(sq.Gt{"boards_history.insert_at": mmModel.GetTimeForMillis(since).Format(time.RFC3339)}).
+		Where(sq.Eq{"boards.team_id": teamID}).
+		Where(sq.Eq{"boards.id": boardIDs}).
+		Where(sq.NotEq{"boards_history.modified_by": "system"}).
+		Where(sq.Eq{"boards.delete_at": 0}).
+		GroupBy("boards.id, boards_history.id, boards_history.modified_by")
+
+	blocksHistoryQuery := s.getQueryBuilder(db).
+		Select("boards.id, boards.icon, boards.title, count(blocks_history.id) as count, blocks_history.modified_by, boards.created_by").
+		Prefix("UNION ALL").
+		From(s.tablePrefix + "blocks_history as blocks_history").
+		Join(s.tablePrefix + "boards as boards on blocks_history.board_id = boards.id").
+		Where(sq.Gt{"blocks_history.insert_at": mmModel.GetTimeForMillis(since).Format(time.RFC3339)}).
+		Where(sq.Eq{"boards.team_id": teamID}).
+		Where(sq.Eq{"boards.id": boardIDs}).
+		Where(sq.NotEq{"blocks_history.modified_by": "system"}).
+		Where(sq.Eq{"boards.delete_at": 0}).
+		GroupBy("boards.id, blocks_history.board_id, blocks_history.modified_by")
+
+	boardsActivity := boardsHistoryQuery.SuffixExpr(blocksHistoryQuery)
+
+	insightsQuery := s.getQueryBuilder(db).Select(
+		fmt.Sprintf("id, title, icon, sum(count) as activity_count, %s as active_users, created_by", s.concatenationSelector("distinct modified_by", ",")),
+	).
+		FromSelect(boardsActivity, "boards_and_blocks_history").
+		GroupBy("id, title, icon, created_by").
+		OrderBy("activity_count desc").
+		Offset(uint64(offset)).
+		Limit(uint64(limit))
```

Comment:
Pagination here looks great but doesn't help the unbounded query of the history table above.  If I were to log in with one of my load test accounts to check results this query could return millions of rows.  @mkraft are you certain we'll address this before enabling in a large instance?

Reply:
Thanks for pointing this out @wiggin77 , genuine concern indeed.

All I could think of is, [this line](https://github.com/mattermost/focalboard/pull/3005/files/194e1c28e9a4d9428f9677de8c7a16574ec7cd69#diff-456b30574cd5964ecde484e5d156d4c8878aa6274bb257588fabc073a9f72538R33) limits the `blocks_history` entries to be in the last `28 days` at maximum. Do you think this information helps reduce the number of rows?

Reply:
That 28 day bound certainly helps.  My load test scenario is a pathological case, however it could also happen if someone imports a large archive.  That said we can revisit this knowing the 28 day bound is there.